synapse/handlers/federation_event.py

```diff
@@ -1168,30 +1190,26 @@ async def get_event(event_id: str) -> None:
                 "Persisting %i of %i remaining events", len(roots), len(event_map)
             )
 
-            await self._auth_and_persist_fetched_events(destination, room_id, roots)
+            await self._auth_and_persist_fetched_events_inner(origin, room_id, roots)
```

Comment:
Hmm, we actually shouldn't need to persist the events separately, e.g. we don't do this when we are persisting the state/auth chains when joining a room?

Reply:
I'm not really following you.

The idea here (and it's an existing idea from https://github.com/matrix-org/synapse/pull/10771, not new in this PR) is that we make sure we auth each event's auth events before the event itself. This is all part of trying to fix https://github.com/matrix-org/synapse/issues/9595. 

And yes, I fear we don't do it on joins, but that's a bug we should also fix.

Reply:
For every event we have persisted we need to have persisted its entire auth chain, but I think its fine to persist the entire auth chain in one go? (We might want to sort the event list topologically to make sure the chunking works in the persist_events class). In the persist events class if we fail to persist one event in a batch then we'll fail the entire batch, so there shouldn't be a danger of us persisting an event but not its auth events?

Reply:
We discussed this in #synapse-dev.

Part of the confusion here is that I'm specifically concerned about event *auth*, rather than persisting the events per se. 

Erik's concern is that this could end up with a lot of round-trips between federation reader and event persister for a long auth chain.

So we can look at optimising this so that we do all the auth up-front, and *then* persist everything together. But I think this will be much easier after some more cleanup here, so we've agreed to park it for now.